### PR TITLE
Feature/post list

### PR DESCRIPTION
**Pull request by fenyolevi**
Implemented post list and post detail components, as well as the dialog for message sending.

### CODE REVIEW

File: src/app/services/post.service.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        )
+    );
+
+    getPosts(): Observable<PostExtended[]> {
+        return this.postsExtended$;
+    };
+
+    getPostById(id: number): Observable<PostExtended | undefined> {
+        return this.postsExtended$.pipe(
+            map(posts => posts.find(
```
Comment (gyumolcs) [inline src/app/services/post.service.ts:53]: What is the expectation of this function?

File: src/app/post-list/post-list.component.scss [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    ul {
+      list-style: none;
+      display: flex;
+      flex-wrap: wrap;
+      gap: 2 * $default-gap;
+      padding: 0;
+
+      mat-card {
```
Comment (gyumolcs) [inline src/app/post-list/post-list.component.scss:27]: We just had a big migration issue with the material theme update as those classes were renamed. It is a better practice to use an own class as an abstraction layer, so if you have to switch to another element or UI library, it would be less painful

File: src/app/post-details/post-details.component.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        const dialogRef = this.dialog.open(MessageDialogComponent, {
+            data: {email: email},
+            autoFocus: 'dialog',
+            ariaModal: true,
+            disableClose: true,
+        });
+        dialogRef.afterClosed()
+            .pipe(takeUntilDestroyed(this.destroyRef))
+            .subscribe(messageSent => {
+                if (messageSent) {
```
Comment (gyumolcs) [inline src/app/post-details/post-details.component.ts:36]: 'if' alone in a subscribe body screams for filter() :)

File: src/app/post-details/post-details.component.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+})
+export class PostDetailsComponent {
+    @Input({required: true}) postDetails$: Observable<PostExtended | undefined> = new Observable<PostExtended>();
+
+    private readonly dialog = inject(MatDialog);
+    private readonly snackBar = inject(MatSnackBar);
+    private readonly destroyRef = inject(DestroyRef);
+
+    openDialog(email: string) {
+        const dialogRef = this.dialog.open(MessageDialogComponent, {
```
Comment (gyumolcs) [inline src/app/post-details/post-details.component.ts:27]: no need for the variable

File: src/app/services/post.service.ts (hunk 1 above)
Comment (fenyolevi) [reply src/app/services/post.service.ts:53]: To return the post in an observable with the given id. Am I doing something wrong here? :D

File: src/app/services/post.service.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                })
+            ).filter(post => post.user !== undefined)
+        )
+    );
+
+    getPosts(): Observable<PostExtended[]> {
+        return this.postsExtended$;
+    };
+
+    getPostById(id: number): Observable<PostExtended | undefined> {
```
[the remark this answers is not in the document]
Comment (fenyolevi) [reply src/app/services/post.service.ts:51]: I changes it to postId

File: src/app/post-details/post-details.component.ts (hunk 4 above)
Comment (gyumolcs) [reply src/app/post-details/post-details.component.ts:27]: How about this?
`        this.dialog.open(MessageDialogComponent, {
            data: {email: email},
            autoFocus: 'dialog',
          ariaModal: true,
          disableClose: true,
        }).afterClosed()
          .pipe(takeUntilDestroyed(this.destroyRef))
          .subscribe(messageSent => {
            if (messageSent) {
              this.showSnackBar()
            }
          });`

Comment (fenyolevi) [reply src/app/post-details/post-details.component.ts:27]: I see, that's better :)

File: src/app/app.component.ts [hunk 6]
```diff
@@ -23,4 +28,6 @@ export class AppComponent {
     prepareRoute(outlet: RouterOutlet): RouterOutlet {
         return outlet && outlet.activatedRouteData && outlet.activatedRouteData['animation'];
     }
+
+    protected readonly PostService = PostService;
```
Comment (gyumolcs) [inline src/app/app.component.ts:32]: Why is this needed?

File: src/app/post-list/post-list.component.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+})
+export class PostListComponent {
+    private readonly postService = inject(PostService);
+    readonly LoaderService = LoaderService;
+    private lastPostId = -1;
+    readonly postList: Signal<PostExtended[] | undefined> = signal([])
+    clickedPost$: Observable<PostExtended | undefined> = new Observable<PostExtended>();
+
+    constructor() {
+        this.postList = toSignal(this.postService.getPosts());
```
Comment (gyumolcs) [inline src/app/post-list/post-list.component.ts:28]: There is already an initial value set to this variable

File: src/app/post-list/post-list.component.html [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+                        </mat-card-header>
+                        <mat-card-content>
+                            Click to see more
+                        </mat-card-content>
+                    </mat-card>
+                </li>
+            </ul>
+        </section>
+        <section>
+            <app-post-details [postDetails$]="clickedPost$"/>
```
Comment (gyumolcs) [inline src/app/post-list/post-list.component.html:20]: This data binding is very strange, the variables are initialized in both components with a new Observable. It is worth to just draw first what you really want to do then design the communication between the components

File: src/app/app.component.ts (hunk 6 above)
Comment (fenyolevi) [reply src/app/app.component.ts:32]: Angular added it because of a previous template reference, I forgot to delete it :)

File: src/app/post-list/post-list.component.ts (hunk 7 above)
Comment (fenyolevi) [reply src/app/post-list/post-list.component.ts:28]: True, I corrected it

File: package.json [hunk 9]
```diff
@@ -10,29 +10,31 @@
...
+    "@angular-devkit/build-angular": "^17.0.6",
+    "@angular/cli": "~17.0.6",
+    "@angular/compiler-cli": "^17.0.6",
     "@types/jasmine": "~4.3.0",
     "jasmine-core": "~4.6.0",
     "karma": "~6.4.0",
     "karma-chrome-launcher": "~3.2.0",
     "karma-coverage": "~2.2.0",
     "karma-jasmine": "~5.1.0",
     "karma-jasmine-html-reporter": "~2.1.0",
```
Comment (gyumolcs) [inline package.json:37]: I recommend setting up some enterprise development tools like ESLint and Prettier. Recently we decided to base upon airbnb: https://www.npmjs.com/package/eslint-config-airbnb

File: package.json [hunk 10]
```diff
@@ -26,8 +27,32 @@
   },
   "devDependencies": {
     "@angular-devkit/build-angular": "^17.0.6",
+    "@angular-eslint/eslint-plugin": "^16.0.0",
```
Comment (gyumolcs) [inline package.json:30]: Lot of Angular-related packages aligned their versioning to the Angular version, this is not an exception :)

File: src/app/error-handling/global-error-handler.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  private _dialog = inject(MatDialog);
+
+  public handleError(error: any): void {
+    if (error instanceof HttpErrorResponse) {
+      console.error(
+        `Backend returned code ${error.status},body was:`,
+        error.error,
+      );
+    }
+    this._zone.run(() =>
```
Comment (gyumolcs) [inline src/app/error-handling/global-error-handler.ts:18]: Why is this has to be in a zone?

Comment (fenyolevi) [reply src/app/error-handling/global-error-handler.ts:18]: The reason was in an article I read: "The opening of the dialog takes place in a callback of zone.run, so that the dialog window can be closed even if the error is thrown outside the ngZone. This is for example the case if an error occurs in a lifecycle hook like the ngOnInit function in a component."

File: package.json (hunk 10 above)
Comment (fenyolevi) [reply package.json:30]: I didn't know that auto-completion works for that as well, nice trick, thank you!

File: src/app/error-handling/global-error-handler.ts (hunk 11 above)
Comment (gyumolcs) [reply src/app/error-handling/global-error-handler.ts:18]: Thanks, this aspect was new to me, information like this might worth a comment

File: src/app/error-handling/error-dialog/error-dialog.component.html [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+        <p *ngIf="data.status">
+            <b>Status Code: {{ data.status }}</b>
+        </p>
+        <p>Something went wrong!</p>
+        <p class="error-message">
+            {{ data.message }}
+        </p>
+    </div>
+    <div mat-dialog-actions align="end">
+        <button mat-raised-button mat-dialog-close color="warn">Close</button>
```
Comment (gyumolcs) [inline src/app/error-handling/error-dialog/error-dialog.component.html:13]: If you are bored, then you might extend this dialog with a retry functionality, so next to this button there can be another one and with that the user could repeat the last operation.

File: src/app/error-handling/http-loading.interceptor.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { HttpEvent, HttpInterceptorFn } from '@angular/common/http';
+import { finalize, Observable } from 'rxjs';
+import { LoaderService } from '../services/loader.service';
+
+export const httpLoadingInterceptor: HttpInterceptorFn = (req, next) => {
+  LoaderService.isLoading.set(true);
```
Comment (gyumolcs) [inline src/app/error-handling/http-loading.interceptor.ts:6]: This solution can't handle multiple parallel requests, here is a good article:
https://netbasal.com/take-a-load-off-your-mind-and-let-loadoff-cleanly-handle-your-loading-indicators-in-angular-d95b6490e19c

**Review by gyumolcs — APPROVED** [review]
Nice work!

File: src/app/error-handling/http-loading.interceptor.ts (hunk 13 above)
Comment (fenyolevi) [reply src/app/error-handling/http-loading.interceptor.ts:6]: I'll try it out later, nice package
